### DISCUSSION TITLE
digital caliper data read rp2040

### DISCUSSION THREAD

**Discussion by metabo7000** [micropython/micropython#15155] [RP2040 / Pico] [answered]
Hello! Could someone show me a sample example of how to read micropython rp2040 read digital caliper data, unfortunately I couldn't find a library or anything like that anywhere!
[https://www.instructables.com/Hacked-Digital-Vernier-Caliper-Using-Arduino/](url)

**Comment by 2dof** [comment]
[here](https://www.instructables.com/Reading-Digital-Callipers-with-an-Arduino-USB/) You have some exapmple.
Note that most of digital calibers use  " the BIN 6 format that represents the position as a single 24-bit stream split into six 4-bit nibbles. One nibble is used to transmit different flags, such as units, minus sign, etc. The other five are used to represent the reading in a binary format, with each nibble representing a number between 0 and 15" ([source](https://www.touchdro.com/resources/scales/capacitive/chinese-digital-calipers.html),, but there are also in BCD format (https://www.robotroom.com/Caliper-Digital-Data-Port.html) , So first You should  find out what type of data format  You got (in last link You will find good explanation and decoding procedure example).
Also keep in mind voltage levels  between RP2040 and caliber.

**Comment by peterhinch** [comment]
Interesting - I have an identical caliper...

The instructables article explains the communication protocol and the level shifting, but I think you'll need to write the code to get the data into the RP2. It looks easy. You need a hard interrupt on the rising edge of the clock. The handler reads the state of the data line and assigns the bit to the least significant bit of the result. With each read the result is shifted left. Once 24 bits have been read transfer is complete and you can interpret the output word in line with the instructable.

**Reply by metabo7000** [reply]
It would be nice to see a sample example of a micropython digital caliper so that it can be studied! I'm still only proficient in Arduino and C++!

**Reply by peterhinch** [reply]
As far as I know nobody has done this. Much as I'm tempted to experiment, I have no application in mind. Worse, I'm far too busy with graphics coding at the moment.

**Reply by GitHubsSilverBullet** [reply]
I'd like to submit an alternative idea.

One doesn't need any level shifters at all. The RP2040 has ADC inputs. One and a  half Volts would result in roughly 1850/4096 or 116/256 (8bit resolution)
The ADC runs at up to 2µs per sample, alternating between two inputs results in 4µs per double-sample.
Run that on the second core and send the resulting data to the first core for whatever one wants to do it.
What else, ehmm, nothing. Done!

**Reply by metabo7000** [reply, marked as answer]
problem Upy commpiler?

Traceback (most recent call last):
  File "<stdin>", line 11
SyntaxError: invalid syntax


`#import RPi.GPIO as GPIO
import time

from machine import Pin

def setup():
#    GPIO.setmode(GPIO.BCM)
#    GPIO.setup(CLOCK_PIN, GPIO.IN)
 #   GPIO.setup(DATA_PIN, GPIO.IN)

CLOCK_PIN = Pin(16, Pin.IN)
DATA_PIN = Pin(17, mode=Pin.IN)

def loop():
    time.sleep(0.001)  
#    while GPIO.input(CLOCK_PIN) == GPIO.LOW:

while True:
if CLOCK_PIN.value()==1:
    time.sleep(0.001)  
    tmpTime = time.time() 
#    while GPIO.input(CLOCK_PIN) == GPIO.HIGH:
while True:
if DATA_PIN.value()==0:
        time.sleep(0.001)  
    if (time.time() - tmpTime) < 0.5:
        return     
    readCaliper() 
    print(" ", end="")
    print(round(result, 3), " in ", end="")
    print(round(result*2.54, 3), " cm", end="") 
    if mm:
        print(" mm")
        time.sleep(0.1)
    else:
        print(" in")
        time.sleep(0.1)

buf = ' '
mm = True
value = 0
sign = 1
inches = 0

def readCaliper():
    global buf, mm, value, sign, inches
    sign = 1
    value = 0
    inches = 0
    
    for i in range(24):
        time.sleep(0.001) 
#        while GPIO.input(CLOCK_PIN) == GPIO.LOW:
while True:
if CLOCK_PIN.value()==1:
            time.sleep(0.001) 
  #      while GPIO.input(CLOCK_PIN) == GPIO.HIGH:
            time.sleep(0.001) 
if DATA_PIN_PIN.value()==0
#        if GPIO.input(DATA_PIN) == GPIO.HIGH:
            if i < 20:
                value |= (1 << i)
            if i == 20:
                sign = -1
            if i == 23:
                inches = 1
    
    if mm:
        result = (value * sign) / 100.0
    else:
        result = (value * sign) / (inches * 2.54)

setup()
while True:
    loop()`

**Reply by metabo7000** [reply]
C++ to Upython....
arduino code....
-------
#define CLOCK_PIN 12
#define DATA_PIN  11


void setup() 
{
  Serial.begin(9600);  
  pinMode(CLOCK_PIN, INPUT);
  pinMode(DATA_PIN, INPUT);
  
}

char buf[20];
unsigned long tmpTime;
int sign;
int inches;
long value;
float result;
bool mm = true;

void loop()
{
  while(digitalRead(CLOCK_PIN)==LOW) {}
  tmpTime=micros();
  while(digitalRead(CLOCK_PIN)==HIGH) {}
  if((micros()-tmpTime)<500) return;
  readCaliper(); 
  buf[0]=' ';
  dtostrf(result,6,3,buf+1); strcat(buf," in ");  
  dtostrf(result*2.54,6,3,buf+1); strcat(buf," cm "); 

  if(mm)
  {
    Serial.print(result); Serial.println(" mm");    
    delay(100);
  }
  else
  {
    Serial.print(result); Serial.println(" in");    
    delay(100);
  }
}
void readCaliper()
{
  sign=1;
  value=0;
  inches=0;
  for(int i=0;i<24;i++) {
    while(digitalRead(CLOCK_PIN)==LOW) {}
    while(digitalRead(CLOCK_PIN)==HIGH) {}
    if(digitalRead(DATA_PIN)==HIGH) {
      if(i<20) value|=(1<<i);
      if(i==20) sign=-1;
      if(i==23) inches=1; 
    }
  }
  if(mm)
  {
    result=(value*sign)/100.0;
  }
  else
  {
  result=(value*sign)/(inches?2000.0:100.0); 
  }
  
}


-------

**Reply by peterhinch** [reply]
Please read [this message](https://github.com/orgs/micropython/discussions/9111) re code formatting.

**Reply by metabo7000** [reply]
> Please read [this message](https://github.com/orgs/micropython/discussions/9111) re code formatting.

The interpretation of the forum insertion code is broken! But do you have time for that?
Never ask questions!

**Reply by scruss** [reply]
No, pasting Python into a webpage is *always* broken. Please read the helpful suggestions.

Your first Python code (the one with the `GPIO.setmode(GPIO.BCM)` in it): where did you get it? It's clearly meant for Python on a Raspberry Pi, not MicroPython

**Reply by metabo7000** [reply]
> No, pasting Python into a webpage is _always_ broken. Please read the helpful suggestions.
> 
> Your first Python code (the one with the `GPIO.setmode(GPIO.BCM)` in it): where did you get it? It's clearly meant for Python on a Raspberry Pi, not MicroPython
Inserting the forum code will delete the end of the commented part #, that's why I put it in text format, insert it!
# GPIO.setmode(GPIO.BCM) Comment!

**Reply by GitHubsSilverBullet** [reply]
Are you unwilling or are you unable to follow the recommendations how to post code snippets?

**Reply by metabo7000** [reply]
> No, pasting Python into a webpage is _always_ broken. Please read the helpful suggestions.
> 
> Your first Python code (the one with the `GPIO.setmode(GPIO.BCM)` in it): where did you get it? It's clearly meant for Python on a Raspberry Pi, not MicroPython

deault code Upython

----

import time

from machine import Pin

def setup():


    CLOCK_PIN = Pin(16, Pin.IN)
    DATA_PIN = Pin(17, Pin.IN)

def loop():
    time.sleep(0.001)  


    while True:
        if CLOCK_PIN.value()==1:
            time.sleep(0.001)  
            tmpTime = time.time() 

    while True:
        if DATA_PIN.value()==0:
            time.sleep(0.001)  
        if (time.time() - tmpTime) < 0.5:
            return     
    readCaliper() 
    print(" ", end="")
    print(round(result, 3), " in ", end="")
    print(round(result*2.54, 3), " cm", end="") 
    if mm:
        print(" mm")
        time.sleep(0.1)
    else:
        print(" in")
        time.sleep(0.1)

    buf = ' '
    mm = True
    value = 0
    sign = 1
    inches = 0

def readCaliper():
    global buf, mm, value, sign, inches
    sign = 1
    value = 0
    inches = 0
    
    for i in range(24):
        time.sleep(0.001) 

    while True:
        if CLOCK_PIN.value()==1:
            time.sleep(0.001) 
 
            time.sleep(0.001) 
        if DATA_PIN.value()==0:

            if i < 20:
                value |= (1 << i)
            if i == 20:
                sign = -1
            if i == 23:
                inches = 1
    
        if mm:
            result = (value * sign) / 100.0
        else:
            result = (value * sign) / (inches * 2.54)

setup()
while True:
    loop()
-------

**Reply by metabo7000** [reply]
> Are you unwilling or are you unable to follow the recommendations how to post code snippets?

What do you think? What happens when I paste the code, the character is mixed up?
Anyone who sees it does not understand the chaos!

**Comment by metabo7000** [comment]
[cal4_3.txt](https://github.com/user-attachments/files/15537706/cal4_3.txt)

**Reply by GitHubsSilverBullet** [reply]
```python

import time

from machine import Pin

def setup():
#    GPIO.setmode(GPIO.BCM)
#    GPIO.setup(CLOCK_PIN, GPIO.IN)
 #   GPIO.setup(DATA_PIN, GPIO.IN)

    CLOCK_PIN = Pin(16, Pin.IN)
    DATA_PIN = Pin(17, Pin.IN)

def loop():
    time.sleep(0.001)  


    while True:
        if CLOCK_PIN.value()==1:
            time.sleep(0.001)  
            tmpTime = time.time() 

    while True:
        if DATA_PIN.value()==0:
            time.sleep(0.001)  
        if (time.time() - tmpTime) < 0.5:
            return     
    readCaliper() 
    print(" ", end="")
    print(round(result, 3), " in ", end="")
    print(round(result*2.54, 3), " cm", end="") 
    if mm:
        print(" mm")
        time.sleep(0.1)
    else:
        print(" in")
        time.sleep(0.1)

    buf = ' '
    mm = True
    value = 0
    sign = 1
    inches = 0

def readCaliper():
    global buf, mm, value, sign, inches
    sign = 1
    value = 0
    inches = 0
    
    for i in range(24):
        time.sleep(0.001) 

    while True:
        if CLOCK_PIN.value()==1:
            time.sleep(0.001) 
 
            time.sleep(0.001) 
        if DATA_PIN.value()==0:

            if i < 20:
                value |= (1 << i)
            if i == 20:
                sign = -1
            if i == 23:
                inches = 1
    
        if mm:
            result = (value * sign) / 100.0
        else:
            result = (value * sign) / (inches * 2.54)

setup()
while True:
    loop()
    

```

**Reply by metabo7000** [reply]
> ```python
> ```python
> import time
> 
> from machine import Pin
> 
> def setup():
> #    GPIO.setmode(GPIO.BCM)
> #    GPIO.setup(CLOCK_PIN, GPIO.IN)
>  #   GPIO.setup(DATA_PIN, GPIO.IN)
> 
>     CLOCK_PIN = Pin(16, Pin.IN)
>     DATA_PIN = Pin(17, Pin.IN)
> 
> def loop():
>     time.sleep(0.001)  
> 
> 
>     while True:
>         if CLOCK_PIN.value()==1:
>             time.sleep(0.001)  
>             tmpTime = time.time() 
> 
>     while True:
>         if DATA_PIN.value()==0:
>             time.sleep(0.001)  
>         if (time.time() - tmpTime) < 0.5:
>             return     
>     readCaliper() 
>     print(" ", end="")
>     print(round(result, 3), " in ", end="")
>     print(round(result*2.54, 3), " cm", end="") 
>     if mm:
>         print(" mm")
>         time.sleep(0.1)
>     else:
>         print(" in")
>         time.sleep(0.1)
> 
>     buf = ' '
>     mm = True
>     value = 0
>     sign = 1
>     inches = 0
> 
> def readCaliper():
>     global buf, mm, value, sign, inches
>     sign = 1
>     value = 0
>     inches = 0
>     
>     for i in range(24):
>         time.sleep(0.001) 
> 
>     while True:
>         if CLOCK_PIN.value()==1:
>             time.sleep(0.001) 
>  
>             time.sleep(0.001) 
>         if DATA_PIN.value()==0:
> 
>             if i < 20:
>                 value |= (1 << i)
>             if i == 20:
>                 sign = -1
>             if i == 23:
>                 inches = 1
>     
>         if mm:
>             result = (value * sign) / 100.0
>         else:
>             result = (value * sign) / (inches * 2.54)
> 
> setup()
> while True:
>     loop()
>     
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
>     
> ```

Traceback (most recent call last):

  File "<stdin>", line 4, in <module>

ImportError: no module named 'machine'

**Reply by GitHubsSilverBullet** [reply]
You DO realize that this is the **Micropython** Discussion group?
You DO realize that this is for a Rasberry Pi **PICO?**

**Reply by scruss** [reply]
>Traceback (most recent call last):
> 
> File "", line 4, in
> 
> ImportError: no module named 'machine'

What board are you running MicroPython on? `from machine import Pin`  should work, but isn't working for you.

**Reply by metabo7000** [reply]
> > Traceback (most recent call last):
> > File "", line 4, in
> > ImportError: no module named 'machine'
> 
> What board are you running MicroPython on? `from machine import Pin` should work, but isn't working for you.

Thonny Rasberry Pi PICO!

**Reply by GitHubsSilverBullet** [reply]
```python
>>> 
MPY: soft reboot
MicroPython v1.23.0 on 2024-06-02; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>> from machine import Pin
>>> Pin(0)
Pin(GPIO0, mode=ALT, pull=PULL_DOWN, alt=31)
>>> 
```

If that doesn't work on your system, then there's something seriously wrong.

**Reply by scruss** [reply]
> Thonny Raspberry Pi PICO!

In Thonny, go to the *Tools* → *Options …* menu, then go to the *Interpreter* tab. It should look something like this:
![Interpreter tab from Thonny, showing MicroPython and connected device selection](https://github.com/micropython/micropython/assets/425706/1e223cc0-dda7-4436-a320-79877b2da2e8)
I've highlighted the important bits in red. If you can't get a "Board in FS mode" option, have you flashed MicroPython on the board? This is how you [add the MicroPython firmware](https://projects.raspberrypi.org/en/projects/getting-started-with-the-pico/3).

In particular, if the Interpreter tab says:
![Local Python 3 Interpreter](https://github.com/micropython/micropython/assets/425706/7a43229b-079b-4e58-b2b2-4463ff8c0265)
the script will fail in the same way you reported:
```
Traceback (most recent call last):
  File "<string>", line 4, in <module>
ModuleNotFoundError: No module named 'machine'
```

**Comment by metabo7000** [comment]
Funny man! :)
